File: libs/document-processor-next/src/runtime/__tests__/scope-runtime-context.test.ts
import { createBlue } from '../../test-support/blue.js';
import { describe, expect, it } from 'vitest';

import { ScopeRuntimeContext } from '../scope-runtime-context.js';

const blue = createBlue();

function nodeFrom(json: unknown) {
  return blue.jsonValueToNode(json);
}

describe('ScopeRuntimeContext', () => {
  it('enqueues and drains triggered queue respecting cut-off', () => {
    const context = new ScopeRuntimeContext('/scope');
    const first = nodeFrom({ id: 1 });
    const second = nodeFrom({ id: 2 });
    const third = nodeFrom({ id: 3 });

    context.enqueueTriggered(first);
    context.enqueueTriggered(second);
    context.markCutOff();
    context.enqueueTriggered(third); // ignored due to cut-off limit

    expect(context.triggeredSize()).toBe(2);
    expect(context.pollTriggered()).toBe(first);
    expect(context.pollTriggered()).toBe(second);
    expect(context.pollTriggered()).toBeUndefined();
  });

  it('records bridgeable events with cut-off limit', () => {
    const context = new ScopeRuntimeContext('/scope');
    const events = [0, 1, 2].map((id) => nodeFrom({ id }));

    context.recordBridgeable(events[0]);
    context.markCutOff();
    context.recordBridgeable(events[1]);
    context.recordBridgeable(events[2]); // ignored

    expect(context.drainBridgeableEvents()).toEqual(events.slice(0, 1));
    expect(context.drainBridgeableEvents()).toEqual([]);
  });

  it('finalizes termination once', () => {
    const context = new ScopeRuntimeContext('/scope');
    context.enqueueTriggered(nodeFrom({ id: 'event' }));
    context.finalizeTermination('GRACEFUL', 'done');

    expect(context.isTerminated()).toBe(true);
    expect(context.terminationKind()).toBe('GRACEFUL');
    expect(context.terminationReason()).toBe('done');
    expect(context.triggeredSize()).toBe(0);
  });
});

File: references/java-sdk/src/test/java/blue/language/processor/ScopeRuntimeContextTest.java
package blue.language.processor;

import blue.language.model.Node;
import org.junit.jupiter.api.Test;

import java.math.BigInteger;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

class ScopeRuntimeContextTest {

    @Test
    void enqueuesAndDrainsTriggeredQueueRespectingCutOff() {
        ScopeRuntimeContext context = new ScopeRuntimeContext("/scope");
        Node first = new Node().properties("id", new Node().value(1));
        Node second = new Node().properties("id", new Node().value(2));
        Node third = new Node().properties("id", new Node().value(3));

        context.enqueueTriggered(first);
        context.enqueueTriggered(second);
        context.markCutOff();
        context.enqueueTriggered(third);

        assertEquals(2, context.triggeredSize());
        assertEquals(first, context.pollTriggered());
        assertEquals(second, context.pollTriggered());
        assertNull(context.pollTriggered());
    }

    @Test
    void recordsBridgeableEventsWithCutOffLimit() {
        ScopeRuntimeContext context = new ScopeRuntimeContext("/scope");
        Node first = new Node().properties("id", new Node().value(0));
        Node second = new Node().properties("id", new Node().value(1));
        Node third = new Node().properties("id", new Node().value(2));

        context.recordBridgeable(first);
        context.markCutOff();
        context.recordBridgeable(second);
        context.recordBridgeable(third);

        List<Node> drained = context.drainBridgeableEvents();
        assertEquals(1, drained.size());
        assertEquals(new BigInteger("0"), drained.get(0).getProperties().get("id").getValue());
        assertTrue(context.drainBridgeableEvents().isEmpty());
    }

    @Test
    void finalizesTerminationOnceAndClearsQueue() {
        ScopeRuntimeContext context = new ScopeRuntimeContext("/scope");
        context.enqueueTriggered(new Node().properties("id", new Node().value("event")));
        context.finalizeTermination(ScopeRuntimeContext.TerminationKind.GRACEFUL, "done");
        context.finalizeTermination(ScopeRuntimeContext.TerminationKind.FATAL, "ignored");

        assertTrue(context.isTerminated());
        assertEquals(ScopeRuntimeContext.TerminationKind.GRACEFUL, context.terminationKind());
        assertEquals("done", context.terminationReason());
        assertEquals(0, context.triggeredSize());
    }
}
